### DISCUSSION TITLE
Dependency initialization for an app, when dependency imported from a swift package.

### DISCUSSION THREAD

**Discussion by cprince-foreflight** [pointfreeco/swift-dependencies#181] [Ideas]
I'm having problems initializing a `@Dependency` for an app or target which consumes a Swift package which defines that `@Dependency`. My specific case is that the `liveValue` has no reasonable default, and that it needs to be overridden by the consuming package. I've looked at the [overriding discussed here](https://swiftpackageindex.com/pointfreeco/swift-dependencies/main/documentation/dependencies/overridingdependencies), but I want the override scoped for the entire application. I think [that overriding](https://swiftpackageindex.com/pointfreeco/swift-dependencies/main/documentation/dependencies/overridingdependencies) doesn't do that.

In the defining Swift package, I'm currently exposing the `liveValue` as a `public static var`, but that doesn't really seem good.

Thoughts?

**Comment by kamcma** [comment]
https://swiftpackageindex.com/pointfreeco/swift-dependencies/main/documentation/dependencies/livepreviewtest#Separating-interface-and-implementation

I think this is the answer. In the module that defines the dependency, you only conform it to `TestDependencyKey`. (Every endpoint can simply be failing. The `@DependencyClient` macro would do this for you automatically.) Then, in the consuming module, you extend the dependency to further conform to `DependencyKey`, where you provide the live value.

Feature modules can depend on only the definition module, but the app entry point module depends on (or perhaps contains) the live implementation, so it will be used by the full app.

**Reply by acosmicflamingo** [reply]
Sounds like what Isowords does for APIClient: 
https://github.com/pointfreeco/isowords/tree/main/Sources/ApiClientLive

https://github.com/pointfreeco/isowords/tree/main/Sources/ApiClient

**Reply by cprince-foreflight** [reply]
Nodding. Thanks! I was initially confused about the `Test` in `TestDependencyKey`. I initially thought that `Test` meant it should **not** be used with code intended for production, but only for code intended for tests.
